File: src/main/java/com/tkr/ingress/IngressClassifier.java
package com.tkr.ingress;

import com.tkr.types.WireTypes;
import com.tkr.types.WireTypes.IngressKind;
import com.tkr.types.WireTypes.Status;
import com.tkr.util.BoundsUtil;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

/**
 * Classifies ingress byte streams into FIX, SWIFT, or TKR wire formats.
 * Ported from legacy-cpp ingress_classifier.cc.
 */
public final class IngressClassifier {

    public static final class IngressProbe {
        public IngressKind kind = IngressKind.UNKNOWN;
        public int magicOrTag;
        public int frameHintBytes;
        public float confidence;
    }

    public static final class IngressClassifyResult {
        public Status status = Status.OK;
        public IngressKind primaryKind = IngressKind.UNKNOWN;
        public boolean isMixedStream;
        public int leadingSkip;
        public List<IngressProbe> probes = new ArrayList<>();
    }

    public static final class IngressClassifierConfig {
        public boolean acceptMixed = true;
        public boolean preferBinaryMagic = true;
        public int minFixHeaderLen = 10;
    }

    private final IngressClassifierConfig config;

    public IngressClassifier() {
        this(new IngressClassifierConfig());
    }

    public IngressClassifier(IngressClassifierConfig config) {
        this.config = config != null ? config : new IngressClassifierConfig();
    }

    public IngressKind classify(byte[] data, int offset, int length) {
        return classifyBinary(data, offset, length).primaryKind;
    }

    public IngressKind classify(byte[] data) {
        return classify(data, 0, data != null ? data.length : 0);
    }

    public IngressClassifyResult classifyBinary(byte[] data, int offset, int length) {
        if (data == null || length == 0) {
            IngressClassifyResult result = new IngressClassifyResult();
            result.status = Status.TRUNCATED;
            return result;
        }
        String view = new String(data, offset, length, StandardCharsets.US_ASCII);
        return classify(view, data, offset, length);
    }

    public IngressClassifyResult classify(String bytes) {
        byte[] data = bytes.getBytes(StandardCharsets.US_ASCII);
        return classify(bytes, data, 0, data.length);
    }

    private IngressClassifyResult classify(String bytes, byte[] data, int offset, int length) {
        IngressClassifyResult result = new IngressClassifyResult();
        result.status = Status.OK;
        result.primaryKind = IngressKind.UNKNOWN;
        if (bytes.isEmpty()) {
            result.status = Status.TRUNCATED;
            return result;
        }
        int printable = countPrintablePrefix(data, offset, length);
        if (printable > 0 && printable < length) result.leadingSkip = printable;

        IngressProbe fixProbe = probeFix44(bytes);
        if (fixProbe.confidence > 0) result.probes.add(fixProbe);

        IngressProbe swiftProbe = probeSwift(bytes);
        if (swiftProbe.confidence > 0) result.probes.add(swiftProbe);

        IngressProbe batchProbe = probeBatchWire(data, offset, length);
        if (batchProbe.confidence > 0) result.probes.add(batchProbe);

        IngressProbe envelopeProbe = probeEnvelopeWire(data, offset, length);
        if (envelopeProbe.confidence > 0) result.probes.add(envelopeProbe);

        IngressProbe sessionProbe = probeSessionWire(data, offset, length);
        if (sessionProbe.confidence > 0) result.probes.add(sessionProbe);

        if (result.probes.isEmpty()) {
            result.status = Status.UNKNOWN_FORMAT;
            return result;
        }

        result.probes.sort(Comparator.comparingDouble((IngressProbe p) -> p.confidence).reversed());
        result.primaryKind = result.probes.get(0).kind;

        if (result.probes.size() >= 2 && config.acceptMixed) {
            float delta = result.probes.get(0).confidence - result.probes.get(1).confidence;
            if (delta < 0.12f) {
                result.isMixedStream = true;
                result.primaryKind = IngressKind.MIXED_STREAM;
            }
        }
        return result;
    }

    public boolean looksLikeFix44(String bytes) {
        return bytes != null && bytes.length() >= 8 && bytes.startsWith("8=FIX.4.4");
    }

    public boolean looksLikeSwiftMt940(String bytes) {
        if (bytes == null) return false;
        return bytes.contains("{1:") || bytes.contains("{4:") || bytes.contains(":20:") || bytes.contains(":61:");
    }

    public boolean looksLikeBatchMagic(int magic) {
        return magic == WireTypes.BATCH_MAGIC;
    }

    public boolean looksLikeEnvelopeMagic(int magic) {
        return magic == WireTypes.ENVELOPE_MAGIC;
    }

    public boolean looksLikeSessionMagic(int magic) {
        return magic == WireTypes.SESSION_MAGIC;
    }

    public int readMagicLe(byte[] data, int offset, int length) {
        if (data == null || length - offset < 4) return 0;
        return BoundsUtil.readU32Le(data, offset);
    }

    public int countPrintablePrefix(byte[] data, int offset, int length) {
        int count = 0;
        for (int i = offset; i < offset + length; i++) {
            char ch = (char) (data[i] & 0xFF);
            if (ch >= 0x20 && ch <= 0x7E || ch == '\r' || ch == '\n' || ch == '\u0001') count++;
            else break;
        }
        return count;
    }

    public IngressProbe probeFix44(String bytes) {
        IngressProbe probe = new IngressProbe();
        probe.kind = IngressKind.FIX44;
        if (bytes.length() < config.minFixHeaderLen || !looksLikeFix44(bytes)) return probe;
        probe.confidence = 0.55f;
        if (bytes.contains("\u00019=")) {
            probe.confidence += 0.20f;
            probe.frameHintBytes = bytes.length();
        }
        if (bytes.contains("\u000135=")) probe.confidence += 0.15f;
        if (bytes.contains("\u000110=")) probe.confidence += 0.10f;
        probe.confidence = clampConfidence(probe.confidence);
        return probe;
    }

    public IngressProbe probeSwift(String bytes) {
        IngressProbe probe = new IngressProbe();
        probe.kind = IngressKind.SWIFT_MT940;
        probe.frameHintBytes = bytes.length();
        if (looksLikeSwiftMt940(bytes)) probe.confidence = 0.50f;
        if (bytes.contains(":61:")) {
            probe.confidence += 0.25f;
            probe.magicOrTag = 61;
        }
        if (bytes.contains(":86:")) probe.confidence += 0.15f;
        if (bytes.contains(":60F:") || bytes.contains(":62F:")) probe.confidence += 0.10f;
        probe.confidence = clampConfidence(probe.confidence);
        return probe;
    }

    public IngressProbe probeBatchWire(byte[] data, int offset, int length) {
        IngressProbe probe = new IngressProbe();
        probe.kind = IngressKind.BATCH_WIRE;
        if (data == null || length - offset < 28) return probe;
        int magic = readMagicLe(data, offset, length);
        probe.magicOrTag = magic;
        if (!looksLikeBatchMagic(magic)) return probe;
        probe.confidence = 0.70f;
        int version = BoundsUtil.readU16Le(data, offset + 4);
        if (version == WireTypes.WIRE_VERSION) probe.confidence += 0.15f;
        int recordCount = readMagicLe(data, offset + 8, length);
        if (recordCount <= WireTypes.MAX_BATCH_RECORDS) {
            probe.confidence += 0.10f;
            probe.frameHintBytes = 28 + recordCount * 32;
        }
        probe.confidence = clampConfidence(probe.confidence);
        return probe;
    }

    public IngressProbe probeEnvelopeWire(byte[] data, int offset, int length) {
        IngressProbe probe = new IngressProbe();
        probe.kind = IngressKind.ENVELOPE_WIRE;
        if (data == null || length - offset < 24) return probe;
        int magic = readMagicLe(data, offset, length);
        probe.magicOrTag = magic;
        if (!looksLikeEnvelopeMagic(magic)) return probe;
        probe.confidence = 0.75f;
        int channelCount = BoundsUtil.readU16Le(data, offset + 6);
        if (channelCount <= WireTypes.MAX_ENVELOPE_CHANNELS) {
            probe.confidence += 0.15f;
            probe.frameHintBytes = 24 + channelCount * 20;
        }
        probe.confidence = clampConfidence(probe.confidence);
        return probe;
    }

    public IngressProbe probeSessionWire(byte[] data, int offset, int length) {
        IngressProbe probe = new IngressProbe();
        probe.kind = IngressKind.SESSION_WIRE;
        if (data == null || length - offset < 24) return probe;
        int magic = readMagicLe(data, offset, length);
        probe.magicOrTag = magic;
        if (!looksLikeSessionMagic(magic)) return probe;
        probe.confidence = 0.72f;
        int legCount = BoundsUtil.readU16Le(data, offset + 6);
        if (legCount <= WireTypes.MAX_SESSION_LEGS) {
            probe.confidence += 0.18f;
            probe.frameHintBytes = 24 + legCount * 28;
        }
        probe.confidence = clampConfidence(probe.confidence);
        return probe;
    }

    public IngressClassifyResult classifyWithComplianceHint(String bytes, boolean requireMarginPath) {
        IngressClassifyResult result = classify(bytes);
        if (result.status != Status.OK || result.probes.isEmpty()) return result;
        for (IngressProbe probe : result.probes) {
            probe.confidence = clampConfidence(probe.confidence + scoreComplianceRelevance(probe, bytes));
        }
        result.probes.sort(Comparator.comparingDouble((IngressProbe p) -> p.confidence).reversed());
        result.primaryKind = result.probes.get(0).kind;
        if (requireMarginPath) {
            boolean marginOk = false;
            for (IngressProbe probe : result.probes) {
                if (probe.kind == IngressKind.SWIFT_MT940 || probe.kind == IngressKind.BATCH_WIRE
                        || probe.kind == IngressKind.ENVELOPE_WIRE) {
                    marginOk = true;
                    break;
                }
            }
            if (!marginOk) result.status = Status.MARGIN_BREACH;
        }
        return result;
    }

    private float scoreComplianceRelevance(IngressProbe probe, String bytes) {
        float bonus = 0f;
        if (probe.kind == IngressKind.FIX44 && probeHasAllocationTags(bytes)) bonus += 0.08f;
        if (probe.kind == IngressKind.SWIFT_MT940 && probeHasMarginSettlementTags(bytes)) bonus += 0.06f;
        if (probe.kind == IngressKind.BATCH_WIRE && probe.magicOrTag == WireTypes.BATCH_MAGIC) bonus += 0.05f;
        return bonus;
    }

    private static boolean probeHasAllocationTags(String bytes) {
        return bytes.contains("\u000178=") || bytes.contains("\u000179=") || bytes.contains("\u0001467=");
    }

    private static boolean probeHasMarginSettlementTags(String bytes) {
        return bytes.contains(":62F:") || bytes.contains(":60F:");
    }

    public float clampConfidence(float value) {
        if (value < 0f) return 0f;
        if (value > 1f) return 1f;
        return value;
    }

    public IngressProbe probe(byte[] data, int offset, int length) {
        IngressClassifyResult result = classifyBinary(data, offset, length);
        return result.probes.isEmpty() ? new IngressProbe() : result.probes.get(0);
    }

    public boolean isBinaryWire(IngressKind kind) {
        return kind == IngressKind.BATCH_WIRE || kind == IngressKind.ENVELOPE_WIRE || kind == IngressKind.SESSION_WIRE;
    }

    public String summarize(IngressClassifyResult result) {
        return "kind=" + result.primaryKind + " mixed=" + result.isMixedStream + " probes=" + result.probes.size();
    }
}

File: legacy-cpp/src/ingress/ingress_classifier.cc
#include "tkr/ingress/ingress_classifier.h"

#include "tkr/util/bounds.h"

#include <algorithm>
#include <cctype>

namespace tkr {
namespace ingress {
namespace {

bool IsPrintableAscii(char ch) {
  return ch >= 0x20 && ch <= 0x7E;
}

bool HasFixPrefix(std::string_view bytes) {
  if (bytes.size() < 8) {
    return false;
  }
  return bytes.substr(0, 8) == "8=FIX.4.4";
}

bool HasSwiftBlockMarker(std::string_view bytes) {
  return util::FindSubstring(bytes, "{1:", 0) != std::string_view::npos ||
         util::FindSubstring(bytes, "{4:", 0) != std::string_view::npos ||
         util::FindSubstring(bytes, ":20:", 0) != std::string_view::npos ||
         util::FindSubstring(bytes, ":61:", 0) != std::string_view::npos;
}

float ClampConfidence(float value) {
  if (value < 0.0f) {
    return 0.0f;
  }
  if (value > 1.0f) {
    return 1.0f;
  }
  return value;
}

}  // namespace

IngressClassifier::IngressClassifier(IngressClassifierConfig config)
    : config_(config) {}

bool IngressClassifier::LooksLikeFix44(std::string_view bytes) {
  return HasFixPrefix(bytes);
}

bool IngressClassifier::LooksLikeSwiftMt940(std::string_view bytes) {
  return HasSwiftBlockMarker(bytes);
}

bool IngressClassifier::LooksLikeBatchMagic(std::uint32_t magic) {
  return magic == kBatchMagic;
}

bool IngressClassifier::LooksLikeEnvelopeMagic(std::uint32_t magic) {
  return magic == kEnvelopeMagic;
}

bool IngressClassifier::LooksLikeSessionMagic(std::uint32_t magic) {
  return magic == kSessionMagic;
}

std::uint32_t IngressClassifier::ReadMagicLe(const std::uint8_t* data,
                                             std::size_t size) const {
  if (data == nullptr || size < 4) {
    return 0;
  }
  return static_cast<std::uint32_t>(data[0]) |
         (static_cast<std::uint32_t>(data[1]) << 8) |
         (static_cast<std::uint32_t>(data[2]) << 16) |
         (static_cast<std::uint32_t>(data[3]) << 24);
}

std::size_t IngressClassifier::CountPrintablePrefix(std::string_view bytes) const {
  std::size_t count = 0;
  for (char ch : bytes) {
    if (!IsPrintableAscii(ch) && ch != '\r' && ch != '\n' && ch != '\001') {
      break;
    }
    ++count;
  }
  return count;
}

IngressProbe IngressClassifier::ProbeFix44(std::string_view bytes) const {
  IngressProbe probe{};
  probe.kind = kIngressFix44;
  probe.magic_or_tag = 0;
  probe.frame_hint_bytes = 0;
  probe.confidence = 0.0f;

  if (bytes.size() < config_.min_fix_header_len) {
    return probe;
  }
  if (!HasFixPrefix(bytes)) {
    return probe;
  }
  probe.confidence = 0.55f;
  const std::size_t body_len_tag = util::FindSubstring(bytes, "\0019=", 0);
  if (body_len_tag != std::string_view::npos) {
    probe.confidence += 0.20f;
    probe.frame_hint_bytes = bytes.size();
  }
  const std::size_t msg_type = util::FindSubstring(bytes, "\00135=", 0);
  if (msg_type != std::string_view::npos) {
    probe.confidence += 0.15f;
  }
  const std::size_t checksum = util::FindSubstring(bytes, "\00110=", 0);
  if (checksum != std::string_view::npos) {
    probe.confidence += 0.10f;
  }
  probe.confidence = ClampConfidence(probe.confidence);
  return probe;
}

IngressProbe IngressClassifier::ProbeSwift(std::string_view bytes) const {
  IngressProbe probe{};
  probe.kind = kIngressSwiftMt940;
  probe.confidence = 0.0f;
  probe.frame_hint_bytes = bytes.size();

  if (HasSwiftBlockMarker(bytes)) {
    probe.confidence = 0.50f;
  }
  if (util::FindSubstring(bytes, ":61:", 0) != std::string_view::npos) {
    probe.confidence += 0.25f;
    probe.magic_or_tag = 61;
  }
  if (util::FindSubstring(bytes, ":86:", 0) != std::string_view::npos) {
    probe.confidence += 0.15f;
  }
  if (util::FindSubstring(bytes, ":60F:", 0) != std::string_view::npos ||
      util::FindSubstring(bytes, ":62F:", 0) != std::string_view::npos) {
    probe.confidence += 0.10f;
  }
  probe.confidence = ClampConfidence(probe.confidence);
  return probe;
}

IngressProbe IngressClassifier::ProbeBatchWire(const std::uint8_t* data,
                                               std::size_t size) const {
  IngressProbe probe{};
  probe.kind = kIngressBatchWire;
  probe.confidence = 0.0f;
  if (data == nullptr || size < 28) {
    return probe;
  }
  const std::uint32_t magic = ReadMagicLe(data, size);
  probe.magic_or_tag = magic;
  if (!LooksLikeBatchMagic(magic)) {
    return probe;
  }
  probe.confidence = 0.70f;
  const std::uint16_t version =
      static_cast<std::uint16_t>(data[4]) |
      (static_cast<std::uint16_t>(data[5]) << 8);
  if (version == kWireVersion) {
    probe.confidence += 0.15f;
  }
  const std::uint32_t record_count = ReadMagicLe(data + 8, size - 8);
  if (record_count <= kMaxBatchRecords) {
    probe.confidence += 0.10f;
    const std::size_t min_frame = 28 + static_cast<std::size_t>(record_count) * 32;
    probe.frame_hint_bytes = min_frame;
  }
  probe.confidence = ClampConfidence(probe.confidence);
  return probe;
}

IngressProbe IngressClassifier::ProbeEnvelopeWire(const std::uint8_t* data,
                                                  std::size_t size) const {
  IngressProbe probe{};
  probe.kind = kIngressEnvelopeWire;
  if (data == nullptr || size < 24) {
    return probe;
  }
  const std::uint32_t magic = ReadMagicLe(data, size);
  probe.magic_or_tag = magic;
  if (!LooksLikeEnvelopeMagic(magic)) {
    return probe;
  }
  probe.confidence = 0.75f;
  const std::uint16_t channel_count =
      static_cast<std::uint16_t>(data[6]) |
      (static_cast<std::uint16_t>(data[7]) << 8);
  if (channel_count <= kMaxEnvelopeChannels) {
    probe.confidence += 0.15f;
    probe.frame_hint_bytes = 24 + static_cast<std::size_t>(channel_count) * 20;
  }
  probe.confidence = ClampConfidence(probe.confidence);
  return probe;
}

IngressProbe IngressClassifier::ProbeSessionWire(const std::uint8_t* data,
                                                 std::size_t size) const {
  IngressProbe probe{};
  probe.kind = kIngressSessionWire;
  if (data == nullptr || size < 24) {
    return probe;
  }
  const std::uint32_t magic = ReadMagicLe(data, size);
  probe.magic_or_tag = magic;
  if (!LooksLikeSessionMagic(magic)) {
    return probe;
  }
  probe.confidence = 0.72f;
  const std::uint16_t leg_count =
      static_cast<std::uint16_t>(data[6]) |
      (static_cast<std::uint16_t>(data[7]) << 8);
  if (leg_count <= kMaxSessionLegs) {
    probe.confidence += 0.18f;
    probe.frame_hint_bytes = 24 + static_cast<std::size_t>(leg_count) * 28;
  }
  probe.confidence = ClampConfidence(probe.confidence);
  return probe;
}

IngressClassifyResult IngressClassifier::Classify(std::string_view bytes) {
  IngressClassifyResult result{};
  result.status = Status::kOk;
  result.primary_kind = kIngressUnknown;
  result.is_mixed_stream = false;
  result.leading_skip = 0;

  if (bytes.empty()) {
    result.status = Status::kTruncated;
    return result;
  }

  const std::size_t printable = CountPrintablePrefix(bytes);
  if (printable > 0 && printable < bytes.size()) {
    result.leading_skip = printable;
  }

  IngressProbe fix_probe = ProbeFix44(bytes);
  if (fix_probe.confidence > 0.0f) {
    result.probes.push_back(fix_probe);
  }

  IngressProbe swift_probe = ProbeSwift(bytes);
  if (swift_probe.confidence > 0.0f) {
    result.probes.push_back(swift_probe);
  }

  const std::uint8_t* bin_data = reinterpret_cast<const std::uint8_t*>(bytes.data());
  IngressProbe batch_probe = ProbeBatchWire(bin_data, bytes.size());
  if (batch_probe.confidence > 0.0f) {
    result.probes.push_back(batch_probe);
  }
  IngressProbe envelope_probe = ProbeEnvelopeWire(bin_data, bytes.size());
  if (envelope_probe.confidence > 0.0f) {
    result.probes.push_back(envelope_probe);
  }
  IngressProbe session_probe = ProbeSessionWire(bin_data, bytes.size());
  if (session_probe.confidence > 0.0f) {
    result.probes.push_back(session_probe);
  }

  if (result.probes.empty()) {
    result.status = Status::kUnknownFormat;
    return result;
  }

  std::sort(result.probes.begin(), result.probes.end(),
            [](const IngressProbe& a, const IngressProbe& b) {
              return a.confidence > b.confidence;
            });

  result.primary_kind = result.probes.front().kind;

  if (result.probes.size() >= 2 && config_.accept_mixed) {
    const float delta = result.probes[0].confidence - result.probes[1].confidence;
    if (delta < 0.12f) {
      result.is_mixed_stream = true;
      result.primary_kind = kIngressMixedStream;
    }
  }

  return result;
}

IngressClassifyResult IngressClassifier::ClassifyBinary(const std::uint8_t* data,
                                                      std::size_t size) {
  if (data == nullptr || size == 0) {
    IngressClassifyResult result{};
    result.status = Status::kTruncated;
    return result;
  }
  std::string_view view(reinterpret_cast<const char*>(data), size);
  return Classify(view);
}

namespace {

bool ProbeHasAllocationTags(std::string_view bytes) {
  return util::FindSubstring(bytes, "\00178=", 0) != std::string_view::npos ||
         util::FindSubstring(bytes, "\00179=", 0) != std::string_view::npos ||
         util::FindSubstring(bytes, "\001467=", 0) != std::string_view::npos;
}

bool ProbeHasMarginSettlementTags(std::string_view bytes) {
  return util::FindSubstring(bytes, ":62F:", 0) != std::string_view::npos ||
         util::FindSubstring(bytes, ":60F:", 0) != std::string_view::npos;
}

float ScoreComplianceRelevance(const IngressProbe& probe, std::string_view bytes) {
  float bonus = 0.0f;
  if (probe.kind == kIngressFix44 && ProbeHasAllocationTags(bytes)) {
    bonus += 0.08f;
  }
  if (probe.kind == kIngressSwiftMt940 && ProbeHasMarginSettlementTags(bytes)) {
    bonus += 0.06f;
  }
  if (probe.kind == kIngressBatchWire &&
      probe.magic_or_tag == kBatchMagic) {
    bonus += 0.05f;
  }
  return bonus;
}

}  // namespace

IngressClassifyResult IngressClassifier::ClassifyWithComplianceHint(
    std::string_view bytes, bool require_margin_path) {
  IngressClassifyResult result = Classify(bytes);
  if (result.status != Status::kOk || result.probes.empty()) {
    return result;
  }

  for (IngressProbe& probe : result.probes) {
    probe.confidence = ClampConfidence(probe.confidence +
                                       ScoreComplianceRelevance(probe, bytes));
  }

  std::sort(result.probes.begin(), result.probes.end(),
            [](const IngressProbe& a, const IngressProbe& b) {
              return a.confidence > b.confidence;
            });
  result.primary_kind = result.probes.front().kind;

  if (require_margin_path) {
    bool margin_ok = false;
    for (const IngressProbe& probe : result.probes) {
      if (probe.kind == kIngressSwiftMt940 ||
          probe.kind == kIngressBatchWire ||
          probe.kind == kIngressEnvelopeWire) {
        margin_ok = true;
        break;
      }
    }
    if (!margin_ok) {
      result.status = Status::kMarginBreach;
    }
  }
  return result;
}

}  // namespace ingress
}  // namespace tkr
